### PR TITLE
[Paddle]Add chapter_attention-mechanisms

### PR DESCRIPTION
**Pull request by tngt**

### CODE REVIEW

File: chapter_attention-mechanisms/bahdanau-attention.md [hunk 1]
```diff
@@ -244,6 +251,57 @@ class Seq2SeqAttentionDecoder(AttentionDecoder):
...
+    def init_state(self, enc_outputs, enc_valid_lens, *args):
+        # outputs的形状为(batch_size，num_steps，num_hiddens).
+        # hidden_state的形状为(num_layers，batch_size，num_hiddens)
+        outputs, hidden_state = enc_outputs
+        return (outputs.transpose((1, 0, 2)), hidden_state, enc_valid_lens)
+
+    def forward(self, X, state):
+        # enc_outputs的形状为(batch_size,num_steps,num_hiddens).
+        # hidden_state的形状为(num_layers,batch_size,
+        # num_hiddens)
```
Comment (cheungdaven) [inline chapter_attention-mechanisms/bahdanau-attention.md:277]: don't start new line here.

File: chapter_attention-mechanisms/nadaraya-waston.md [hunk 2]
```diff
@@ -331,6 +389,13 @@ values = tf.reshape(tf.range(20.0), shape = (2, 10))
 tf.matmul(tf.expand_dims(weights, axis=1), tf.expand_dims(values, axis=-1)).numpy()
 ```
 
+```{.python .input}
+#@tab paddle
+weights = paddle.ones((2, 10)) * 0.1
+values = paddle.arange(20, dtype='float32').reshape((2,10))
```
Comment (cheungdaven) [inline chapter_attention-mechanisms/nadaraya-waston.md:395]: (2, 10)

File: chapter_attention-mechanisms/nadaraya-waston.md [hunk 3]
```diff
@@ -485,6 +579,22 @@ for epoch in range(5):
     animator.add(epoch + 1, float(loss))
 ```
 
+```{.python .input}
+#@tab paddle
+net = NWKernelRegression()
+loss = nn.MSELoss(reduction='none')
+trainer = paddle.optimizer.SGD(learning_rate=0.5, parameters=net.parameters())
+animator = d2l.Animator(xlabel='epoch', ylabel='loss', xlim=[1,5])
```
Comment (cheungdaven) [inline chapter_attention-mechanisms/nadaraya-waston.md:587]: [1, 5]

**Review by cheungdaven — CHANGES_REQUESTED** [review]

File: chapter_attention-mechanisms/nadaraya-waston.md (hunk 2 above)
Comment (w5688414) [reply chapter_attention-mechanisms/nadaraya-waston.md:395]: Done

File: chapter_attention-mechanisms/nadaraya-waston.md (hunk 3 above)
Comment (w5688414) [reply chapter_attention-mechanisms/nadaraya-waston.md:587]: Done

File: chapter_attention-mechanisms/bahdanau-attention.md (hunk 1 above)
Comment (w5688414) [reply chapter_attention-mechanisms/bahdanau-attention.md:277]: Done

File: chapter_attention-mechanisms/attention-scoring-functions.md [hunk 4]
```diff
@@ -137,6 +144,26 @@ def masked_softmax(X, valid_lens):
...
+    # X:3D张量，valid_lens:1D或2D张量
+    if valid_lens is None:
+        return nn.functional.softmax(X, axis=-1)
+    else:
+        shape = X.shape
+        if valid_lens.dim() == 1:
+            valid_lens = paddle.repeat_interleave(valid_lens, shape[1])
+        else:
+            valid_lens = valid_lens.reshape((-1,))
+    #     # 最后一轴上被掩蔽的元素使用一个非常大的负值替换，从而其softmax输出为0
```
[the remark this answers is not in the document]
Comment (w5688414) [reply chapter_attention-mechanisms/attention-scoring-functions.md:161]: Done

**Comment by cheungdaven** [comment]
@w5688414 could you fix the CI issue? It seems that showheatmap function is not in the paddle.py file.

**Comment by w5688414** [comment]
> map function is not in the paddle.py file.

I updated the paddle.py, it works fine again, can you review this chapter again?

**Review by cheungdaven — APPROVED** [review]

**Pull request merged**